modules/hooks.nix

```diff
@@ -2874,12 +2894,15 @@ lib.escapeShellArgs (lib.concatMap (ext: [ "--ghc-opt" "-X${ext}" ]) hooks.ormol
             builtins.toString script;
           files = "\\.nix$";
         };
-      nixfmt =
+      # nixfmt was renamed to nixfmt-classic.
+      # The hook has been deprecated to free up the name for when the new RFC-style nixfmt becomes stable.
+      nixfmt = nixfmt-classic;
+      nixfmt-classic =
         {
-          name = "nixfmt";
-          description = "Nix code prettifier.";
+          name = "nixfmt-classic";
+          description = "Nix code prettifier (classic).";
           package = tools.nixfmt-classic;
-          entry = "${hooks.nixfmt.package}/bin/nixfmt ${lib.optionalString (hooks.nixfmt.settings.width != null) "--width=${toString hooks.nixfmt.settings.width}"}";
+          entry = "${hooks.nixfmt-classic.package}/bin/nixfmt ${lib.optionalString (hooks.nixfmt-classic.settings.width != null) "--width=${toString hooks.nixfmt-classic.settings.width}"}";
```

Comment:
This line lead to a [regression](https://github.com/nix-community/nixvim/pull/2188#issuecomment-2336471966) in nixvim, where we were using `nixfmt.enable`, but preemptively setting `nixfmt.package = pkgs.nixfmt-rfc-style`.

Because the `nixfmt` hook is now a copy of `nixfmt-classic` (but not a proper two-way alias), this line is ignoring any config under `nixfmt.*` and only considering `nixfmt-classic.*`.

I think this could be solved by using `mkAliasOptionModule` or `mkRenamedOptionModule` instead of doing a simple copy on line 2899.

E.g:

```nix
imports = [
  (lib.mkAliasOptionModule [ "hooks" "nixfmt" ] [ "hooks" "nixfmt-classic" ])
];
```

This way any configuration defined under the `hooks.nixfmt` option will be forwarded to the `nixfmt-classic` option (and vice-versa).

Reply:
Thanks for the feedback, and sorry for causing trouble!

Unfortunately, we can't use `mkAliasOptionModule` or `mkRenamedOptionModule` because `hooks` is a freeform set of submodule. See https://github.com/NixOS/nixpkgs/issues/96006.

We'll need to add some machinery to handle this properly.

Reply:
Thanks for the added context!

I believe that issue only relates to the warnings produced by `mkRenamedOptionModule` not being able to "escape" the submodule. If I understand correctly, `mkAliasOptionModule` would still work as intended.

You could of course do things a bit more manually, for example with `mkAliasAndWrapDefinitions`, although if `hooks` is a submodule I think that'd be easier to do _within_ the submodule:

```nix
{ lib, config, options, ... }:
{
  config = lib.mkIf config.nixfmt.enable {
    nixfmt-classic = lib.mkAliasAndWrapDefinitions lib.id options.nixfmt;

    # Or if that is recursive:
    nixfmt-classic = builtins.removeAttrs (lib.mkAliasAndWrapDefinitions lib.id options.nixfmt) [ "enable" ];
  }
}
```

But I think `mkAliasOptionModule` should be fine :thinking:

Reply:
You're absolutely right. I just realised we've been trying to apply the renames to the top-level options, which is why the `toRename` machinery would always fail to evaluate. Once applied to the submodule, it works as excepted.